src/OpenTelemetry.Api/Context/Propagation/TraceContextFormat.cs

```diff
@@ -74,18 +74,18 @@ public bool IsInjected<T>(T carrier, Func<T, string, IEnumerable<string>> getter
         }
 
         /// <inheritdoc/>
-        public ActivityContext Extract<T>(T carrier, Func<T, string, IEnumerable<string>> getter)
+        public ActivityContext Extract<T>(ActivityContext activityContext, T carrier, Func<T, string, IEnumerable<string>> getter)
```

Comment:
Not sure i understood this.. An existing context is passed.. We extract and create a new context and return it.. so why would we need the existing context? Its currently used as returned context when nothing can be extracted.. I was imagining more like - we build on the given activity context.. So the returned one has all traceid,spanid,tracestate it extracted, along with whatever tracestate already present in the passed activitycontext.

I want to hear expert thoughts here @reyang @CodeBlanch

Reply:
For other classes that implement the ITextFormat, the activityContext passed would be the default, so, if something happen, for example, the carrier is null or other validation, we would return default. So, nothing changed.

For the case where we have the CompositePropagator, so, the first time, that would be the default, and, from that to the next one, it will keep passing and reutilizing the activitycontext. With that, the carrier would have in the end all the data from all the ITextFormat.

Below, the link from OT.python: https://github.com/open-telemetry/opentelemetry-python/blob/004896f1987e84b2d8f5c84bb1505099e178694c/opentelemetry-api/src/opentelemetry/propagators/composite.py#L23
Below, the link from OT.java: https://github.com/open-telemetry/opentelemetry-java/blob/690651b65a938c903c35d65aef3182999fcb1b19/context_prop/src/main/java/io/opentelemetry/context/propagation/DefaultContextPropagators.java#L142-L147

Reply:
If
SystemA extractor gives me context with IDs and TraceState as "systemAField=something"
SystemB extractor gives me context with IDs and TraceState as "systemBField=somethingelse" 

The if I use a composite propagator cosisting of "SystemA", "SystemB", should I get the context with IDs from SystemB (as its added after A?), and TraceState as combining both "systemAField=something,systemBField=somethingelse"

Is the above correct? The spec seems vague about this. Hopefully Reiley/Mikel have more context on context.

Reply:
This was a tough one! When I first looked at @eddynaka's implementation, it was like this:

```csharp
        public ActivityContext Extract<T>(T carrier, Func<T, string, IEnumerable<string>> getter)
        {
            foreach (var textFormat in this.textFormats)
            {
                activityContext = textFormat.Extract(carrier, getter);
            }

            return activityContext;
        }
```

Loop through the child propagators, extract from them. But what if the first one succeeds, returns a valid context, and then all others fail? You end up with no context.

Initially, we thought the spec was bugged. But looking at py and java, we realized we were missing the `context` parameter on the `Extract` "base" operation: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/api-propagators.md#extract

The current version is actually how the spec says to do it, and how py and java are doing it! It seems OK to me. The goal is to extract the context. What `CompositePropagator` is doing is allowing to you inject in multiple formats when you transmit. Presumably to support mixing in with multiple downstream systems? When you extract it, you really only need one version of the context. All the others should be duplicates, so the loop really could return on the first valid one it finds?

But agree the spec is a bit strange and vague definitely for propagation. We should try to get it improved, and hammer out how the `IsInjected` should work (https://github.com/open-telemetry/opentelemetry-specification/issues/674).

Reply:
674 got tagged for after-ga, it won't likely be resolved this year :-:(

So if the 1st Extract is a custompropogator which looked at "customhttpheader" and extract a Context, we return that right? If that custompropogator is not extracting the W3C `tracestate` header, then we are likely losing that information unless every extractor is given a chance to run and cumulatively build `tracestate`. 

Also even if any number of extractors are run, they should either not be able to obtain traceid/spanid, or obtain traceid/span which would match others extractors who are able to successfully extract. -- is this correct?

If I am wrong in my tracestate interpretation and propagation, I'd love to get it clarified. Hopefully then then spec can be made more clear. Today spec simply says , for "tracestate" , refer to W3C spec.

Reply:
Hello :wave:

When using a composite propagator, the first successfully extracted context should be returned, ignoring the remaining propagators (if any). This is to allow contexts to be extracted from different formats, in a multiformat setup but, as @CodeBlanch said, if multiple formats are received, they should contain the same information.

Reply:
@MikeGoldsmith , based on that, should we apply the same strategy to IsInjected? If we have one that is injected, so, return true?

Reply:
just updated. Let me know what do u think

Reply:
After reading more times, I come to think that this is correct.